File: packages/dotdog/src/grammar.ts
// .dog DSL Grammar — formal definition of the spec genome language
//
// A .dog file is a structured document with:
//   - Sections delimited by ## or ### headings
//   - Prose (markdown paragraphs, tables, lists)
//   - Structured blocks (entity, relationship, event, prediction definitions)
//
// Grammar (EBNF-like):
//
//   dog-file      ::= section*
//   section       ::= heading block*
//   heading       ::= "##" text | "###" text
//   block         ::= prose-block | entity-block | relationship-block | event-block | prediction-block
//   prose-block   ::= paragraph | table | code-fence | list
//   entity-block  ::= entity-heading (prose)? yaml-body
//   relationship-block ::= relationship-heading (prose)? yaml-body
//   event-block   ::= event-heading (prose)? yaml-body
//   entity-heading    ::= "###" "Entity:" entity-name
//   relationship-heading ::= "###" "Relationship:" source "→" target
//   event-heading     ::= "###" "Event:" event-name
//   yaml-body    ::= "```" yaml-content "```"
//   entity-name  ::= identifier
//   source       ::= identifier
//   target       ::= identifier
//   identifier   ::= [A-Za-z_][A-Za-z0-9_]*
//
// YAML schema within entity blocks:
//   entity:       name
//   type:         node | entity | relationship | event | task | prediction | vector
//   description:  string
//   properties:
//     prop-name: {type: string, required: boolean, constraints: string?, default: any?}
//   states:       [state, state, ...]
//   lifecycle:    state → state → state (→ state)*
//
// YAML schema within relationship blocks:
//   relationship: source → target
//   verb:         contains | depends_on | references | implements | calls | owns | precedes
//   cardinality:  1:1 | 1:N | N:1 | N:M
//   required:     boolean
//   cascade:      none | delete | nullify | restrict
//   invariants:   string

// --- AST Node Types ---

export type ASTNode =
  | DocumentNode
  | SectionNode
  | EntityNode
  | RelationshipNode
  | EventNode
  | PredictionNode
  | EndpointNode
  | ProseNode
  | TableNode;


// Parser error with line context
export interface ParseError {
  message: string;
  line?: number;
  context?: string; // the problematic line or section heading
}

export interface DocumentNode {
  kind: 'document';
  sections: SectionNode[];
  errors: ParseError[];
}

export interface SectionNode {
  kind: 'section';
  level: 2 | 3;
  heading: string;
  blocks: BlockNode[];
  lineStart: number;
  lineEnd: number;
}

export type BlockNode = EntityNode | RelationshipNode | EventNode | PredictionNode | EndpointNode | ProseNode | TableNode;

export interface EntityNode {
  kind: 'entity';
  name: string;
  description: string;
  type: string;
  properties: Record<string, PropertyDef>;
  states: string[];
  lifecycle: string[];
  yaml: Record<string, unknown>;
  lineStart: number;
  lineEnd: number;
}

export interface PropertyDef {
  type: string;
  required: boolean;
  default?: unknown;
  constraints?: string;
  example?: string;
}

export interface RelationshipNode {
  kind: 'relationship';
  source: string;
  target: string;
  verb: string;
  cardinality: string;
  required: boolean;
  cascade: string;
  invariants: string[];
  yaml: Record<string, unknown>;
  lineStart: number;
  lineEnd: number;
}

export interface EventNode {
  kind: 'event';
  name: string;
  trigger: string;
  payload: Record<string, string>;
  preconditions: string[];
  postconditions: string[];
  sideEffects: string[];
  probability: number | null;
  yaml: Record<string, unknown>;
  lineStart: number;
  lineEnd: number;
}

export interface PredictionNode {
  kind: 'prediction';
  statement: string;
  trigger: string;
  timeframe: string;
  confidence: number;
  measurement: string;
  yaml: Record<string, unknown>;
  lineStart: number;
  lineEnd: number;
}

export interface ProseNode {
  kind: 'prose';
  content: string;
  lineStart: number;
  lineEnd: number;
}

export interface TableNode {
  kind: 'table';
  headers: string[];
  rows: string[][];
  lineStart: number;
  lineEnd: number;
}

export interface EndpointNode {
  kind: 'endpoint';
  name: string;
  url: string;
  backup_url?: string;
  method: string;
  expect_status: number;
  expect_body: Record<string, unknown> | null;
  timeout: number;
  yaml: Record<string, unknown>;
  lineStart: number;
  lineEnd: number;
}

File: crates/dotdog/src/grammar.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub type YamlMap = BTreeMap<String, Value>;
pub type EntityView<'a> = (
    &'a str,
    &'a str,
    &'a str,
    &'a BTreeMap<String, PropertyDef>,
    &'a [String],
);

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ParseError {
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Document {
    pub kind: String,
    pub sections: Vec<Section>,
    pub errors: Vec<ParseError>,
}

impl Default for Document {
    fn default() -> Self {
        Self {
            kind: "document".into(),
            sections: Vec::new(),
            errors: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Section {
    pub kind: String,
    pub level: u8,
    pub heading: String,
    pub blocks: Vec<Block>,
    pub line_start: usize,
    pub line_end: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PropertyDef {
    #[serde(rename = "type")]
    pub property_type: String,
    pub required: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub default: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub constraints: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub example: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum Block {
    #[serde(rename_all = "camelCase")]
    Entity {
        name: String,
        description: String,
        #[serde(rename = "type")]
        entity_type: String,
        properties: BTreeMap<String, PropertyDef>,
        states: Vec<String>,
        lifecycle: Vec<String>,
        yaml: YamlMap,
        line_start: usize,
        line_end: usize,
    },
    #[serde(rename_all = "camelCase")]
    Relationship {
        source: String,
        target: String,
        verb: String,
        description: String,
        cardinality: String,
        required: bool,
        cascade: String,
        invariants: Vec<String>,
        yaml: YamlMap,
        line_start: usize,
        line_end: usize,
    },
    #[serde(rename_all = "camelCase")]
    Event {
        name: String,
        trigger: String,
        payload: BTreeMap<String, String>,
        preconditions: Vec<String>,
        postconditions: Vec<String>,
        side_effects: Vec<String>,
        probability: Option<f64>,
        yaml: YamlMap,
        line_start: usize,
        line_end: usize,
    },
    #[serde(rename_all = "camelCase")]
    Prediction {
        statement: String,
        description: String,
        trigger: String,
        timeframe: String,
        confidence: f64,
        measurement: String,
        status: String,
        yaml: YamlMap,
        line_start: usize,
        line_end: usize,
    },
    #[serde(rename_all = "camelCase")]
    Endpoint {
        name: String,
        url: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        backup_url: Option<String>,
        method: String,
        expect_status: u16,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        expect_body: Option<Value>,
        timeout: u64,
        yaml: YamlMap,
        line_start: usize,
        line_end: usize,
    },
    #[serde(rename_all = "camelCase")]
    Prose {
        content: String,
        line_start: usize,
        line_end: usize,
    },
    #[serde(rename_all = "camelCase")]
    Table {
        headers: Vec<String>,
        rows: Vec<Vec<String>>,
        line_start: usize,
        line_end: usize,
    },
}

impl Block {
    pub fn entity_name(&self) -> Option<&str> {
        match self {
            Self::Entity { name, .. } => Some(name),
            Self::Event { name, .. } => Some(name),
            Self::Endpoint { name, .. } => Some(name),
            _ => None,
        }
    }

    pub fn as_entity(&self) -> Option<EntityView<'_>> {
        match self {
            Self::Entity {
                name,
                description,
                entity_type,
                properties,
                states,
                ..
            } => Some((name, description, entity_type, properties, states)),
            _ => None,
        }
    }
}

pub type DocumentNode = Document;
pub type SectionNode = Section;
pub type BlockNode = Block;
